include/inviwo/core/processors/poolprocessor.h

```diff
@@ -0,0 +1,505 @@
+
+    template <typename Result, typename Job>
+    std::shared_ptr<std::packaged_task<Result()>> makeTask(Job&& job, pool::Stop stop,
+                                                           pool::Progress progress);
+
+    template <typename Result, typename Done>
+    static void callDone(InviwoApplication* app,
+                         std::shared_ptr<pool::detail::StateTemplate<Result, Done>> state);
+
+    pool::Options options_;
+    std::shared_ptr<pool::detail::Wrapper> wrapper_;
+    std::vector<std::shared_ptr<pool::detail::State>> states_;
+    std::vector<Submission> queue_;
+    Delay delay_;
+};
+
+namespace pool::detail {
+struct IVW_CORE_API Wrapper {
+    Wrapper(PoolProcessor& pp) : processor(pp) {}
+    PoolProcessor& processor;
+};
+
+struct IVW_CORE_API State {
+    State(std::weak_ptr<Wrapper> processor, size_t count)
+        : processor(processor), count{count}, stop{false}, progress(count) {}
+
+    std::weak_ptr<Wrapper> processor;
+    std::atomic<size_t> count;
+    std::atomic<bool> stop;
+    std::vector<std::atomic<float>> progress;
+    std::future<void> progressUpdate;
+
+    Stop getStop() { return Stop(stop); }
+
+    void setProgress(size_t id, float progress);
+
+    Progress getProgress(size_t id) {
+        IVW_ASSERT(id < count, "Id has to be less then count to be valid");
+        return Progress{*this, id};
+    }
+};
+
+template <typename Result, typename Done>
+struct StateTemplate : State {
+    StateTemplate(std::weak_ptr<Wrapper> processor, size_t count, Done&& done)
+        : State(processor, count), futures{}, done{std::forward<Done>(done)} {}
+
+    std::vector<std::future<Result>> futures;
+    Done done;
+};
+
+template <typename Job>
+struct JobTraits {
+    static_assert(std::is_invocable_v<Job> || std::is_invocable_v<Job, pool::Stop> ||
+                  std::is_invocable_v<Job, Progress> || std::is_invocable_v<Job, Stop, Progress> ||
+                  std::is_invocable_v<Job, Progress, Stop>);
+
+    using CallTest =
+        std::tuple<std::is_invocable<Job>, std::is_invocable<Job, Stop>,
+                   std::is_invocable<Job, Progress>, std::is_invocable<Job, Stop, Progress>,
+                   std::is_invocable<Job, Progress, Stop>>;
+
+    using ValidArgs = std::tuple<std::tuple<>, std::tuple<Stop>, std::tuple<Progress>,
+                                 std::tuple<Stop, Progress>, std::tuple<Progress, Stop>>;
+
+    static constexpr size_t argIndex = util::index_of_derived<std::true_type, CallTest>();
+    using Args = std::tuple_element_t<argIndex, ValidArgs>;
+    using Result = decltype(std::apply(std::declval<Job>(), std::declval<Args>()));
+};
+
+}  // namespace pool::detail
+
+template <typename Result, typename Done>
+inline void PoolProcessor::callDone(
+    InviwoApplication* app, std::shared_ptr<pool::detail::StateTemplate<Result, Done>> state) {
+    static const auto done = [](PoolProcessor& p, auto state) {
```

Comment:
is this lambda actually used? The lambda in dispatchFrontAndForget does no capture it.

Reply:
No need to capture it since it is static